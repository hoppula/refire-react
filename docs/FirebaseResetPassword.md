# FirebaseResetPassword

> Higher order component for Firebase password reset

## Options

**validator** *(input) => Boolean*

Optional. Custom validator function.

```js
import { FirebaseResetPassword } from 'refire-react'
class ResetPasswordComponent extends Component {
  render() {
    const {
      email,
      submit,
      updateEmail,
      validInput,
      error,
      processing,
      completed
    } = this.props
  }
}
export default FirebaseResetPassword({
  validator: optionalCustomValidatorFn
})(ResetPasswordComponent)
```
