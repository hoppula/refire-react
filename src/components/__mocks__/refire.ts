import {Â firebaseActions as actions } from "refire"
export { syncFirebase, firebaseToProps, firebaseReducer } from "refire"
export const firebaseActions = {
  ...actions,
  passwordLogin: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  createUser: jest.fn((email, password) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  resetPassword: jest.fn((email) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  write: jest.fn((method, path, value, ownProps) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
  clearWriteErrors: jest.fn((path) => {
    return (dispatch, getState) => {
      return Promise.resolve()
    }
  }),
}
