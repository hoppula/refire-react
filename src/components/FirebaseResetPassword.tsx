import * as React from "react"
import { Component } from "react"
import * as ReactRedux from "react-redux"
import { connect } from "react-redux"
import { firebaseActions, firebaseToProps } from "refire"

const { resetPassword, clearResetPasswordError } = firebaseActions

const defaultValidator = (input: string): boolean => {
  return !!input
}

export interface Status {
  errors: {
    resetPassword: string
  },
  processing: {
    resetPassword: boolean
  },
  completed: {
    resetPassword: boolean
  }
}

export interface ResetPasswordOptions {
  validator?: (email: string) => boolean
}

export interface ResetPasswordProps {
  readonly _status?: Status,
  readonly dispatch?: ReactRedux.Dispatch<{}>,
  readonly children?: React.ReactNode,
  readonly context?: any
}

interface ResetPasswordState {
  readonly email: string
}

export interface ExtraProps {
  email?: string,
  submit?: (event: React.FormEvent<HTMLElement>) => void,
  updateEmail?: (event: React.FormEvent<HTMLInputElement>) => void,
  validInput?: boolean,
  error?: string,
  processing?: boolean,
  completed?: boolean
}

export default function(options: ResetPasswordOptions = {}) {

  const { validator = defaultValidator } = options

  return function <P>(
    WrappedComponent: ReactRedux.Component<P>
  ) {

    class FirebaseResetPassword extends Component<ResetPasswordProps & P, ResetPasswordState> {

      constructor(props: ResetPasswordProps & P) {
        super(props)
        this.state = {
          email: null
        }
      }

      submit = (event: React.FormEvent<HTMLElement>) => {
        event.preventDefault()
        this.props.dispatch(
          resetPassword(this.state.email)
        )
      }

      updateEmail = (event: React.FormEvent<HTMLInputElement>) => {
        const { errors: { resetPassword: error } } = this.props._status
        if (error) {
          this.props.dispatch(clearResetPasswordError())
        }
        this.setState({ email: event.currentTarget.value })
      }

      render() {
        const validInput: boolean = validator(this.state.email)
        const {
          errors: { resetPassword: error },
          processing: { resetPassword: processing },
          completed: { resetPassword: completed }
        } = this.props._status

        const extraProps: ExtraProps = {
          email: this.state.email,
          submit: this.submit,
          updateEmail: this.updateEmail,
          validInput,
          error,
          processing,
          completed
        }

        return (
          <WrappedComponent
            { ...this.props }
            { ...extraProps }
          />
        )
      }

    }

    return connect<ResetPasswordProps, void, P>(
      firebaseToProps(["_status"])
    )(FirebaseResetPassword)
  }
}
