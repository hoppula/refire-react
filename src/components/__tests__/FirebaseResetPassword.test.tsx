import * as React from "react"
import { Provider } from "react-redux"
import * as renderer from "react-test-renderer"
import { applyMiddleware, compose, createStore } from "redux"
import thunk from "redux-thunk"

import { firebaseActions } from "refire"
// resetPassword action is mocked in ../__mocks__/refire.ts
const { resetPassword } = firebaseActions

import FirebaseResetPassword from "../FirebaseResetPassword"

function createEmptyStore() {
  return compose(
    applyMiddleware(thunk)
  )(createStore)((state, action) => {
    return {
      firebase: {
        initialValuesReceived: true,
        errors: {
          resetPassword: null
        },
        processing: {
          resetPassword: false
        },
        completed: {
          resetPassword: false
        }
      }
    }
  })
}

describe("FirebaseResetPassword", () => {
  it("updateEmail updates email", () => {

    interface EmailProps {
      email?: string,
      updateEmail?: (event: any) => void,
    }
    function Email({ email, updateEmail }: EmailProps) {
      return (
        <div onClick={ updateEmail }>
          { email }
        </div>
      )
    }

    const ResetPasswordComponent = FirebaseResetPassword({})(Email)

    let component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <ResetPasswordComponent />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({currentTarget: {value: "test@test.com"}})

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("submit calls dispatch prop", () => {

    interface SubmitProps {
      dispatch?: () => void,
      submit?: (event: any) => void
    }
    function Submit({ submit }: SubmitProps) {
      return (
        <button onClick={ submit }>
          Submit
        </button>
      )
    }

    const SubmitComponent = FirebaseResetPassword({})(Submit)

    let component = renderer.create(
      <Provider store={ createEmptyStore() }>
        <SubmitComponent />
      </Provider>
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.onClick({ preventDefault() {} })
    expect(resetPassword.mock.calls.length).toBe(1)
  })

  it("validator functions properly")
  it("errors are generated")
  it("processing prop gets updated")
  it("completed prop gets updated")

})
