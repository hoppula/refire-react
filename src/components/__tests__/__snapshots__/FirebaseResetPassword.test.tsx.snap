// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirebaseResetPassword submit calls dispatch prop 1`] = `
<button
  onClick={[Function]}
>
  Submit
</button>
`;

exports[`FirebaseResetPassword updateEmail updates email 1`] = `
<div
  onClick={[Function]}
/>
`;

exports[`FirebaseResetPassword updateEmail updates email 2`] = `
<div
  onClick={[Function]}
>
  test@test.com
</div>
`;
